File: frontend/src/types/stage.ts
export interface Repo {
  id: string;
  repo_name: string;
  repo_url: string;
  default_branch: string;
  enabled: boolean;
  global_pre_script: string | null;
  global_pre_script_scope: string;
  global_pre_script_lock_enabled: boolean;
  global_pre_script_lock_key: string | null;
  global_pre_script_lock_timeout_secs: number;
  global_post_script: string | null;
  global_post_script_scope: string;
  global_post_script_lock_enabled: boolean;
  global_post_script_lock_key: string | null;
  global_post_script_lock_timeout_secs: number;
  created_at: string;
  updated_at: string;
}

export interface StageConfig {
  id: string;
  repo_id: string;
  stage_name: string;
  command: string | null;
  command_mode: string;
  required_cpu: number;
  required_memory_mb: number;
  required_disk_mb: number;
  max_duration_secs: number;
  execution_order: number;
  parallel_group: string | null;
  allow_worker_migration: boolean;
  job_type: string;
  created_at: string;
  updated_at: string;
}

export interface StageScript {
  id: string;
  stage_config_id: string;
  worker_id: string | null;
  script_type: 'pre' | 'post';
  script_scope: 'worker' | 'master';
  script: string;
  lock_enabled: boolean;
  lock_key: string | null;
  lock_timeout_secs: number;
  created_at: string;
  updated_at: string;
}

export interface CreateRepoRequest {
  repo_name: string;
  repo_url: string;
  default_branch?: string;
  enabled?: boolean;
  global_pre_script?: string | null;
  global_pre_script_scope?: string;
  global_pre_script_lock_enabled?: boolean;
  global_pre_script_lock_key?: string | null;
  global_pre_script_lock_timeout_secs?: number;
  global_post_script?: string | null;
  global_post_script_scope?: string;
  global_post_script_lock_enabled?: boolean;
  global_post_script_lock_key?: string | null;
  global_post_script_lock_timeout_secs?: number;
}

export interface Webhook {
  id: string;
  repo_id: string;
  provider: string;
  secret: string;
  events: string[];
  enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface WebhookDelivery {
  id: string;
  webhook_id: string;
  event: string;
  status_code: number;
  response_time_ms: number;
  delivered_at: string;
  success: boolean;
}

export interface CreateStageConfigRequest {
  stage_name: string;
  command?: string | null;
  command_mode?: string;
  required_cpu?: number;
  required_memory_mb?: number;
  required_disk_mb?: number;
  max_duration_secs?: number;
  execution_order?: number;
  parallel_group?: string;
  job_type?: string;
}

File: crates/ci-core/src/models/stage.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;
use uuid::Uuid;

/// Repository configuration
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct Repo {
    pub id: Uuid,
    pub repo_name: String,
    pub repo_url: String,
    pub default_branch: String,
    pub enabled: bool,
    pub max_concurrent_builds: i32,
    pub cancel_superseded: bool,
    pub global_pre_script: Option<String>,
    pub global_pre_script_scope: String,
    pub global_post_script: Option<String>,
    pub global_post_script_scope: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Stage configuration for a repo
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct StageConfig {
    pub id: Uuid,
    pub repo_id: Uuid,
    pub stage_name: String,
    pub command: Option<String>,
    pub required_cpu: i32,
    pub required_memory_mb: i32,
    pub required_disk_mb: i32,
    pub max_duration_secs: i32,
    pub execution_order: i32,
    pub parallel_group: Option<String>,
    pub allow_worker_migration: bool,
    pub job_type: String,
    pub depends_on: Vec<String>,
    pub required_labels: Vec<String>,
    pub max_retries: i32,
    /// Command mode: "fixed" (default), "optional" (user can override), "required" (user must provide)
    pub command_mode: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Pre/post scripts for a stage
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct StageScript {
    pub id: Uuid,
    pub stage_config_id: Uuid,
    pub worker_id: Option<String>,
    pub script_type: String,  // "pre" or "post"
    pub script_scope: String, // "worker" or "master"
    pub script: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Webhook configuration for a repo
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct Webhook {
    pub id: Uuid,
    pub repo_id: Uuid,
    pub provider: String,
    pub secret: String,
    pub events: Vec<String>,
    pub enabled: bool,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Worker reservation record
#[derive(Debug, Clone, Serialize, Deserialize, ToSchema)]
pub struct WorkerReservation {
    pub id: Uuid,
    pub worker_id: String,
    pub job_group_id: Uuid,
    pub reserved_at: DateTime<Utc>,
    pub expires_at: DateTime<Utc>,
    pub released_at: Option<DateTime<Utc>>,
    pub release_reason: Option<String>,
}
